### ISSUE TITLE
First-class custom themes

### ISSUE THREAD

**Issue by wadetregaskis** [KevinBatdorf/code-block-pro#270]
The theme customisation is pretty useless for me as it stands because it cripples the syntax analysis, at least for languages like Swift.  e.g. it then confuses numeric literals and type names, can't understand macro and property wrapper arguments all (they default back to plain text), no longer distinguishes argument labels from their values, etc.

With at least some of the built-in themes, the syntax analysis is far more powerful (much better than other engines I've tried, including Prism.js, Highlight.js, Enlighter, etc).  But I can't really do anything useful with them re. CSS customisation because the individual elements (`span`s) don't have any distinguishing classes.

Thus there doesn't appear to be any way to define a 'proper' custom theme, one that has access to the full syntax grammar.

**Comment by KevinBatdorf** [comment]
Yeah it's quite limited. One option would be to forcefully override the styles. Find a theme that tokenizes in a way you like, then target it by the actual color:

```css
span[style*="#96E072"] {
  color:red !important;
}
```

I'm not sure when I'll have a better solution, so that's probably the best option at the moment.

**Comment by wadetregaskis** [comment]
Having looked more into the underpinnings (Shiki etc), I think what I actually want is simply the ability to use an existing VS Code theme that doesn't happen to be one of the ones built in to the plug-in.  e.g. https://github.com/vbaswin/Xcode14-VScode-Extension, or perhaps older variants like https://github.com/lunacookies/xcode-theme or https://github.com/smockle-archive/xcode-default-theme (I can't be sure which will actually work best without trying them - which I guess I could do by downloading VS Code etc…).

I did consider the approach you mentioned - I'd already been doing that a little bit to adjust fonts - but it seems inherently pretty fragile - I'd first have to find a theme which works well with Swift, from the shortlisted options, _and_ has distinct colours for all the different types I need, and then hope that it never changes any of them.

Would it be possible to support using an arbitrary theme file?  I did glance through the source code in case there were some obvious point I could hack in with PHP and add to the themes list, but nothing jumped out at me.  I'm okay with something a little hacky - e.g. using functions.php - if that's what it takes.

**Comment by wadetregaskis** [comment]
Sorry for the delay in following up.  This led me down quite a rabbit-hole.

Upon testing pretty much every existing VS Code "Xcode" theme I could find (in the Marketplace), more than a dozen of them, I discovered that they're all pretty terrible.  VS Code's underlying Swift support is a bit mediocre (at least, without [bespoke extensions](https://marketplace.visualstudio.com/items?itemName=sswg.swift-lang) that probably aren't applicable to a WordPress plug-in like yours).  Yet even so, most of the themes I tested don't even hit the grammar parser's limits because they're little more than existing themes - intended for JavaScript & CSS, predominately - _very slightly_ modified.  They cover some of the more obvious syntax elements, such that they look like they work at very first glance, but they crumble completely under any real scrutiny.

I spent several days working on a new theme to properly support Swift, but - while I did quickly get further than any existing themes - I eventually realised it was going to be a lot more work to finish it, let-alone to get it published into the VS Code Marketplace, and then support it indefinitely.  Creating VS Code themes is laborious and error-prone, because of the complete lack of tooling and infrastructure to support it (i.e. there's no way to even know what the matchable grammar is, short of reverse-engineering the TextMate grammar files).

So, for now at least, I think there's nothing for you to do here.  If you want to keep this bug open, I can keep an eye on developments in the relevant grammar and theme files, and can report here if there's any notable improvements.  There is actually [some work being done in that area as we speak](https://forums.swift.org/t/updated-syntax-highlighting-in-github-vs-code/68972), for the first time in _years_, so there is maybe still hope for this.

**Comment by KevinBatdorf** [comment]
That's surprising and unfortunate. It's interesting that people are still talking about it too. Yeah let's leave this issue open and check back on the issue you linked to to see if some progress has been made.
